servers/rendering/renderer_rd/effects/vrs.cpp

```diff
@@ -0,0 +1,171 @@
+#include "../uniform_set_cache_rd.h"
+#include "servers/xr_server.h"
+
+using namespace RendererRD;
+
+VRS::VRS() {
+	{
+		Vector<String> vrs_modes;
+		vrs_modes.push_back("\n"); // VRS_DEFAULT
+		vrs_modes.push_back("\n#define MULTIVIEW\n"); // VRS_MULTIVIEW
+
+		vrs_shader.shader.initialize(vrs_modes);
+
+		if (!RendererCompositorRD::singleton->is_xr_enabled()) {
+			vrs_shader.shader.set_variant_enabled(VRS_MULTIVIEW, false);
+		}
+
+		vrs_shader.shader_version = vrs_shader.shader.version_create();
+
+		//use additive
+
+		for (int i = 0; i < VRS_MAX; i++) {
+			if (vrs_shader.shader.is_variant_enabled(i)) {
+				vrs_shader.pipelines[i].setup(vrs_shader.shader.version_get_shader(vrs_shader.shader_version, i), RD::RENDER_PRIMITIVE_TRIANGLES, RD::PipelineRasterizationState(), RD::PipelineMultisampleState(), RD::PipelineDepthStencilState(), RD::PipelineColorBlendState::create_disabled(), 0);
+			} else {
+				vrs_shader.pipelines[i].clear();
+			}
+		}
+	}
+}
+
+VRS::~VRS() {
+	vrs_shader.shader.version_free(vrs_shader.shader_version);
+}
+
+void VRS::copy_vrs(RID p_source_rd_texture, RID p_dest_framebuffer, bool p_multiview) {
+	UniformSetCacheRD *uniform_set_cache = UniformSetCacheRD::get_singleton();
+	ERR_FAIL_NULL(uniform_set_cache);
+	MaterialStorage *material_storage = MaterialStorage::get_singleton();
+	ERR_FAIL_NULL(material_storage);
+
+	// setup our uniforms
+	RID default_sampler = material_storage->sampler_rd_get_default(RS::CANVAS_ITEM_TEXTURE_FILTER_NEAREST, RS::CANVAS_ITEM_TEXTURE_REPEAT_DISABLED);
+
+	RD::Uniform u_source_rd_texture(RD::UNIFORM_TYPE_SAMPLER_WITH_TEXTURE, 0, Vector<RID>({ default_sampler, p_source_rd_texture }));
+
+	VRSMode mode = p_multiview ? VRS_MULTIVIEW : VRS_DEFAULT;
+
+	RID shader = vrs_shader.shader.version_get_shader(vrs_shader.shader_version, mode);
+	ERR_FAIL_COND(shader.is_null());
+
+	RD::DrawListID draw_list = RD::get_singleton()->draw_list_begin(p_dest_framebuffer, RD::INITIAL_ACTION_KEEP, RD::FINAL_ACTION_READ, RD::INITIAL_ACTION_KEEP, RD::FINAL_ACTION_DISCARD, Vector<Color>());
+	RD::get_singleton()->draw_list_bind_render_pipeline(draw_list, vrs_shader.pipelines[mode].get_render_pipeline(RD::INVALID_ID, RD::get_singleton()->framebuffer_get_format(p_dest_framebuffer)));
+	RD::get_singleton()->draw_list_bind_uniform_set(draw_list, uniform_set_cache->get_cache(shader, 0, u_source_rd_texture), 0);
+	RD::get_singleton()->draw_list_bind_index_array(draw_list, material_storage->get_quad_index_array());
+	// RD::get_singleton()->draw_list_set_push_constant(draw_list, &vrs_shader.push_constant, sizeof(VRSPushConstant));
+	RD::get_singleton()->draw_list_draw(draw_list, true);
+	RD::get_singleton()->draw_list_end();
+}
+
+void VRS::create_vrs_texture(const int p_base_width, const int p_base_height, const uint32_t p_view_count, RID &p_vrs_texture, RID &p_vrs_fb) {
+	// TODO find a way to skip this if VRS is not supported, but we don't have access to VulkanContext here, even though we're in vulkan.. hmmm
```

Comment:
Did you end up adding feature flags to the RenderingDevice? I imagine that is what you would want to use here to check whether VRS is supported from the RenderingDevice API, that way there is no need to access Vulkan Context

Reply:
I had forgotten about that, thanks for the reminder @clayjohn , I probably won't have time for it today but I'll change that over tomorrow.